modules/react/common/lib/utils/components.ts

```diff
@@ -239,49 +239,51 @@ export function useModelContext<T>(context: React.Context<T>, model?: T): T {
  * }
  */
-export function composeHooks<M, P extends {}, O1 extends {}, O2 extends {}>(
-  hook1: (model: M, props: P) => O1,
-  hook2: (model: M, props: P) => O2
-): (model: M, props: P) => P & O1 & O2;
-export function composeHooks<M, P extends {}, O1 extends {}, O2 extends {}, O3 extends {}>(
-  hook1: (model: M, props: P) => O1,
-  hook2: (model: M, props: P) => O2,
-  hook3: (model: M, props: P) => O3
-): (model: M, props: P) => P & O1 & O2 & O3;
+export function composeHooks<M, R, P extends {}, O1 extends {}, O2 extends {}>(
+  hook1: (model: M, props: P, ref: React.Ref<R>) => O1,
+  hook2: (model: M, props: P, ref: React.Ref<R>) => O2
+): (model: M, props: P, ref: React.Ref<R>) => P & O1 & O2;
+export function composeHooks<M, R, P extends {}, O1 extends {}, O2 extends {}, O3 extends {}>(
+  hook1: (model: M, props: P, ref: React.Ref<R>) => O1,
+  hook2: (model: M, props: P, ref: React.Ref<R>) => O2,
+  hook3: (model: M, props: P, ref: React.Ref<R>) => O3
+): (model: M, props: P, ref: React.Ref<R>) => P & O1 & O2 & O3;
 export function composeHooks<
   M,
+  R,
   P extends {},
   O1 extends {},
   O2 extends {},
   O3 extends {},
   O4 extends {}
 >(
-  hook1: (model: M, props: P) => O1,
-  hook2: (model: M, props: P) => O2,
-  hook3: (model: M, props: P) => O3,
-  hook4: (model: M, props: P) => O4
-): (model: M, props: P) => P & O1 & O2 & O3 & O4;
+  hook1: (model: M, props: P, ref: React.Ref<R>) => O1,
+  hook2: (model: M, props: P, ref: React.Ref<R>) => O2,
+  hook3: (model: M, props: P, ref: React.Ref<R>) => O3,
+  hook4: (model: M, props: P, ref: React.Ref<R>) => O4
+): (model: M, props: P, ref: React.Ref<R>) => P & O1 & O2 & O3 & O4;
 export function composeHooks<
   M,
+  R,
   P extends {},
   O1 extends {},
   O2 extends {},
   O3 extends {},
   O4 extends {},
   O5 extends {}
 >(
-  hook1: (model: M, props: P) => O1,
-  hook2: (model: M, props: P) => O2,
-  hook3: (model: M, props: P) => O3,
-  hook4: (model: M, props: P) => O4,
-  hook5: (model: M, props: P) => O5
-): (model: M, props: P) => P & O1 & O2 & O3 & O4 & O5;
-export function composeHooks<M, P extends {}, O extends {}>(
-  ...hooks: ((model: M, props: P) => O)[]
-): (model: M, props: P) => O {
-  return (model: M, props: P) => {
+  hook1: (model: M, props: P, ref: React.Ref<R>) => O1,
+  hook2: (model: M, props: P, ref: React.Ref<R>) => O2,
+  hook3: (model: M, props: P, ref: React.Ref<R>) => O3,
+  hook4: (model: M, props: P, ref: React.Ref<R>) => O4,
+  hook5: (model: M, props: P, ref: React.Ref<R>) => O5
+): (model: M, props: P, ref: React.Ref<R>) => P & O1 & O2 & O3 & O4 & O5;
+export function composeHooks<M, R, P extends {}, O extends {}>(
+  ...hooks: ((model: M, props: P, ref: React.Ref<R>) => O)[]
+): (model: M, props: P, ref: React.Ref<R>) => O {
+  return (model: M, props: P, ref: React.Ref<R>) => {
     return hooks.reverse().reduce((props: any, hook) => {
-      return hook(model, props);
+      return hook(model, props, props.ref || ref);
```

Comment:
@alanbsmith Here's how `ref` is reduced when composing hooks. Refs are first passed to functions. If the hook returns `ref` in the prop list, it will be used to continue passing down, otherwise it falls back to `ref` passed to composition.